sdks/java/io/amqp/src/main/java/org/apache/beam/sdk/io/amqp/AmqpIO.java

```diff
@@ -0,0 +1,471 @@
+      // they bind the listener
+      List<UnboundedAmqpSource> sources = new ArrayList<>();
+      if (desiredNumSplits > 0) {
+        for (int i = 0; i < desiredNumSplits; i++) {
+          sources.add(new UnboundedAmqpSource(spec));
+        }
+      } else {
+        sources.add(new UnboundedAmqpSource(spec));
+      }
+      return sources;
+    }
+
+    @Override
+    public UnboundedReader<String> createReader(PipelineOptions pipelineOptions,
+                                                AmqpCheckpointMark checkpointMark) {
+      return new UnboundedAmqpReader(this, checkpointMark);
+    }
+
+    @Override
+    public Coder<String> getDefaultOutputCoder() {
+      return StringUtf8Coder.of();
+    }
+
+    @Override
+    public Coder<AmqpCheckpointMark> getCheckpointMarkCoder() {
+      return SerializableCoder.of(AmqpCheckpointMark.class);
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+  }
+
+  private static class UnboundedAmqpReader extends UnboundedSource.UnboundedReader<String> {
+
+    private final UnboundedAmqpSource source;
+
+    private Messenger messenger;
+    private String current;
+    private Instant currentTimestamp;
+    private AmqpCheckpointMark checkpointMark;
+
+    public UnboundedAmqpReader(UnboundedAmqpSource source, AmqpCheckpointMark checkpointMark) {
+      this.source = source;
+      this.current = null;
+      if (checkpointMark != null) {
+        this.checkpointMark = checkpointMark;
+      } else {
+        this.checkpointMark = new AmqpCheckpointMark();
+      }
+    }
+
+    @Override
+    public Instant getWatermark() {
+      return checkpointMark.watermark;
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return currentTimestamp;
+    }
+
+    @Override
+    public String getCurrent() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public UnboundedSource.CheckpointMark getCheckpointMark() {
+      return checkpointMark;
+    }
+
+    @Override
+    public UnboundedAmqpSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      Read spec = source.spec;
+      messenger = Messenger.Factory.create();
+      messenger.start();
+      for (String address : spec.addresses()) {
+        messenger.subscribe(address);
+      }
+      checkpointMark.messenger = messenger;
+      return advance();
+    }
+
+    @Override
+    public boolean advance() {
+      messenger.recv();
+      if (messenger.incoming() > 0) {
+        Message message = messenger.get();
+        Tracker tracker = messenger.incomingTracker();
+        checkpointMark.trackers.add(tracker);
+        currentTimestamp = new Instant(message.getCreationTime());
+        if (currentTimestamp.isBefore(checkpointMark.watermark)) {
```

Comment:
Hm, why? This will set the watermark to the lowest message timestamp ever observed, right?